File: 3Operators/python/operator.py
# ========================================
# Python Operators - Complete Demo
# ========================================

# 1. Arithmetic Operators
print("=== Arithmetic Operators ===")
a, b = 10, 3
print(f"{a} + {b} = {a + b}")      # Addition
print(f"{a} - {b} = {a - b}")      # Subtraction
print(f"{a} * {b} = {a * b}")      # Multiplication
print(f"{a} / {b} = {a / b}")      # Division (float)
print(f"{a} // {b} = {a // b}")    # Floor Division
print(f"{a} % {b} = {a % b}")      # Modulus
print(f"{a} ** {b} = {a ** b}")    # Exponentiation
print()

# 2. Relational (Comparison) Operators
print("=== Relational Operators ===")
x, y = 5, 7
print(f"{x} == {y} → {x == y}")
print(f"{x} != {y} → {x != y}")
print(f"{x} > {y} → {x > y}")
print(f"{x} < {y} → {x < y}")
print(f"{x} >= {y} → {x >= y}")
print(f"{x} <= {y} → {x <= y}")
print()

# 3. Logical Operators
print("=== Logical Operators ===")
p = True
q = False
print(f"p and q → {p and q}")
print(f"p or q  → {p or q}")
print(f"not p    → {not p}")
print(f"(5 > 3) and (10 < 20) → {(5 > 3) and (10 < 20)}")
print()

# 4. Assignment Operators
print("=== Assignment Operators ===")
num = 10
print(f"Initial num = {num}")

num += 5    # num = num + 5
print(f"After += 5  → {num}")

num -= 3
print(f"After -= 3  → {num}")

num *= 2
print(f"After *= 2  → {num}")

num /= 4
print(f"After /= 4  → {num}")

num //= 2
print(f"After //= 2 → {num}")

num **= 3
print(f"After **= 3 → {num}")
print()

# 5. Bitwise Operators
print("=== Bitwise Operators ===")
m, n = 5, 3        # 5 = 101, 3 = 011
print(f"m = {m} (binary: {bin(m)})")
print(f"n = {n} (binary: {bin(n)})")

print(f"m & n  = {m & n}  → {bin(m & n)}")   # AND
print(f"m | n  = {m | n}  → {bin(m | n)}")   # OR
print(f"m ^ n  = {m ^ n}  → {bin(m ^ n)}")   # XOR
print(f"~m     = {~m}     → {bin(~m)}")      # NOT
print(f"m << 2 = {m << 2} → {bin(m << 2)}")   # Left shift
print(f"m >> 1 = {m >> 1} → {bin(m >> 1)}")   # Right shift
print()

# 6. Ternary (Conditional) Operator
print("=== Ternary Operator ===")
age = 20
status = "Adult" if age >= 18 else "Minor"
print(f"Age {age} → {status}")

score = 85
grade = "A" if score >= 90 else "B" if score >= 80 else "C"
print(f"Score {score} → Grade: {grade}")
print()

# 7. Operator Precedence Example
print("=== Operator Precedence Demo ===")
exp1 = 2 + 3 * 4 ** 2          # ** > * > +
print(f"2 + 3 * 4 ** 2 = {exp1}")   # 50

exp2 = (2 + 3) * 4 ** 2
print(f"(2 + 3) * 4 ** 2 = {exp2}") # 80

exp3 = 10 > 5 and 3 < 8 or not 4 == 4
print(f"10 > 5 and 3 < 8 or not 4 == 4 → {exp3}")
print()

# Bonus: No ++ or -- in Python!
print("=== Increment / Decrement ===")
count = 0
count += 1   # This is the Python way
print(f"count after += 1 → {count}")

File: 3Operators/c++/operator.cpp
#include <iostream>
#include <bitset>   
using namespace std;

int main() {
    cout << "=== C++ Operators Demo ===\n\n";

    int a = 10, b = 3;

    cout << "a = " << a << ", b = " << b << "\n\n";

    // 1. Arithmetic Operators
    cout << "1. Arithmetic Operators:\n";
    cout << "a + b  = " << a + b << endl;   // 13
    cout << "a - b  = " << a - b << endl;   // 7
    cout << "a * b  = " << a * b << endl;   // 30
    cout << "a / b  = " << a / b << endl;   // 3  (integer division)
    cout << "a % b  = " << a % b << endl;   // 1
    cout << endl;

    // 2. Relational Operators
    cout << "2. Relational Operators (returns 1 for true, 0 for false):\n";
    cout << "a > b  = " << (a > b) << endl;   // 1
    cout << "a < b  = " << (a < b) << endl;   // 0
    cout << "a >= b = " << (a >= b) << endl;  // 1
    cout << "a <= b = " << (a <= b) << endl;  // 0
    cout << "a == b = " << (a == b) << endl;  // 0
    cout << "a != b = " << (a != b) << endl;  // 1
    cout << endl;

    // 3. Logical Operators
    cout << "3. Logical Operators:\n";
    cout << "(a > 5) && (b < 5) = " << ((a > 5) && (b < 5)) << endl;  // 1
    cout << "(a < 5) || (b < 5) = " << ((a < 5) || (b < 5)) << endl;  // 1
    cout << "!(a == 10)          = " << !(a == 10) << endl;          // 0
    cout << endl;

    // 4. Increment / Decrement
    cout << "4. Increment & Decrement:\n";
    int x = 5;
    cout << "x = " << x << endl;
    cout << "x++   = " << x++ << "  → x becomes " << x << endl;   // post-increment
    cout << "++x   = " << ++x << "  → x becomes " << x << endl;   // pre-increment
    cout << "x--   = " << x-- << "  → x becomes " << x << endl;
    cout << "--x   = " << --x << "  → x becomes " << x << endl;
    cout << endl;

    // 5. Assignment Operators
    cout << "5. Assignment / Compound Assignment Operators:\n";
    int num = 20;
    cout << "num = " << num << endl;
    num += 5;   cout << "num += 5  → " << num << endl;
    num -= 3;   cout << "num -= 3 → " << num << endl;
    num *= 2;   cout << "num *= 2 → " << num << endl;
    num /= 4;   cout << "num /= 4 → " << num << endl;
    num %= 5;   cout << "num %= 5 → " << num << endl;
    cout << endl;

    // 6. Bitwise Operators
    cout << "6. Bitwise Operators (a=10=1010b, b=3=0011b):\n";
    cout << "a & b  = " << (a & b) << "  (" << bitset<8>(a & b) << ")" << endl;  // 2
    cout << "a | b  = " << (a | b) << "  (" << bitset<8>(a | b) << ")" << endl;   // 11
    cout << "a ^ b  = " << (a ^ b) << "  (" << bitset<8>(a ^ b) << ")" << endl;  // 9
    cout << "~a      = " << (~a) << "  (" << bitset<8>(~a) << ")" << endl;       // -11 (two's complement)
    cout << "a << 1 = " << (a << 1) << "  (" << bitset<8>(a << 1) << ")" << endl; // 20
    cout << "a >> 1 = " << (a >> 1) << "  (" << bitset<8>(a >> 1) << ")" << endl; // 5
    cout << endl;

    // 7. Ternary (Conditional) Operator
    cout << "7. Ternary Operator:\n";
    int age = 25;
    string status = (age >= 18) ? "Adult" : "Minor";
    cout << "Age = " << age << " → " << status << endl;

    int max = (a > b) ? a : b;
    cout << "Maximum of a and b = " << max << endl << endl;

    cout << "=== End of Demo ===\n";
    return 0;
}